File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/buffers/StreamAwareQueue.java
/*
 * Copyright (c) 2023 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.integrations.destination_async.buffers;

import io.airbyte.integrations.destination_async.partial_messages.PartialAirbyteMessage;
import java.time.Instant;
import java.util.Optional;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicReference;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public class StreamAwareQueue {

  private final AtomicReference<Instant> timeOfLastMessage;

  private final MemoryBoundedLinkedBlockingQueue<MessageWithMeta> memoryAwareQueue;

  public StreamAwareQueue(final long maxMemoryUsage) {
    memoryAwareQueue = new MemoryBoundedLinkedBlockingQueue<>(maxMemoryUsage);
    timeOfLastMessage = new AtomicReference<>();
  }

  public long getCurrentMemoryUsage() {
    return memoryAwareQueue.getCurrentMemoryUsage();
  }

  public void addMaxMemory(final long maxMemoryUsage) {
    memoryAwareQueue.addMaxMemory(maxMemoryUsage);
  }

  public Optional<Instant> getTimeOfLastMessage() {
    // if the queue is empty, the time of last message is irrelevant
    if (size() == 0) {
      return Optional.empty();
    }
    return Optional.ofNullable(timeOfLastMessage.get());
  }

  public Optional<MemoryBoundedLinkedBlockingQueue.MemoryItem<MessageWithMeta>> peek() {
    return Optional.ofNullable(memoryAwareQueue.peek());
  }

  public int size() {
    return memoryAwareQueue.size();
  }

  public boolean offer(final PartialAirbyteMessage message, final long messageSizeInBytes, final long stateId) {
    if (memoryAwareQueue.offer(new MessageWithMeta(message, stateId), messageSizeInBytes)) {
      timeOfLastMessage.set(Instant.now());
      return true;
    } else {
      return false;
    }
  }

  public MemoryBoundedLinkedBlockingQueue.MemoryItem<MessageWithMeta> take() throws InterruptedException {
    return memoryAwareQueue.take();
  }

  public MemoryBoundedLinkedBlockingQueue.MemoryItem<MessageWithMeta> poll() {
    return memoryAwareQueue.poll();
  }

  public MemoryBoundedLinkedBlockingQueue.MemoryItem<MessageWithMeta> poll(final long timeout, final TimeUnit unit) throws InterruptedException {
    return memoryAwareQueue.poll(timeout, unit);
  }

  public record MessageWithMeta(PartialAirbyteMessage message, long stateId) {}

}

File: airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/integrations/destination/async/buffers/StreamAwareQueue.kt
/*
 * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.cdk.integrations.destination.async.buffers

import io.airbyte.cdk.integrations.destination.async.model.PartialAirbyteMessage
import java.time.Instant
import java.util.Optional
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicReference

class StreamAwareQueue(maxMemoryUsage: Long) {
    private val timeOfLastMessage: AtomicReference<Instant> = AtomicReference()

    private val memoryAwareQueue: MemoryBoundedLinkedBlockingQueue<MessageWithMeta> =
        MemoryBoundedLinkedBlockingQueue(maxMemoryUsage)

    val currentMemoryUsage: Long
        get() = memoryAwareQueue.currentMemoryUsage

    val maxMemoryUsage: Long
        get() = memoryAwareQueue.maxMemoryUsage

    fun addMaxMemory(maxMemoryUsage: Long) {
        memoryAwareQueue.addMaxMemory(maxMemoryUsage)
    }

    val isEmpty: Boolean
        get() = memoryAwareQueue.size() == 0

    fun getTimeOfLastMessage(): Optional<Instant> {
        // if the queue is empty, the time of last message is irrelevant
        if (size() == 0) {
            return Optional.empty()
        }
        return Optional.ofNullable(timeOfLastMessage.get())
    }

    fun peek(): Optional<MemoryBoundedLinkedBlockingQueue.MemoryItem<MessageWithMeta>> {
        return Optional.ofNullable(memoryAwareQueue.peek())
    }

    fun size(): Int {
        return memoryAwareQueue.size()
    }

    fun offer(
        message: PartialAirbyteMessage,
        messageSizeInBytes: Long,
        stateId: Long,
    ): Boolean {
        if (memoryAwareQueue.offer(MessageWithMeta(message, stateId), messageSizeInBytes)) {
            timeOfLastMessage.set(Instant.now())
            return true
        } else {
            return false
        }
    }

    @Throws(InterruptedException::class)
    fun take(): MemoryBoundedLinkedBlockingQueue.MemoryItem<MessageWithMeta> {
        return memoryAwareQueue.take()
    }

    fun poll(): MemoryBoundedLinkedBlockingQueue.MemoryItem<MessageWithMeta>? {
        return memoryAwareQueue.poll()
    }

    @Throws(InterruptedException::class)
    fun poll(
        timeout: Long,
        unit: TimeUnit,
    ): MemoryBoundedLinkedBlockingQueue.MemoryItem<MessageWithMeta>? {
        return memoryAwareQueue.poll(timeout, unit)
    }

    @JvmRecord data class MessageWithMeta(val message: PartialAirbyteMessage, val stateId: Long)
}
